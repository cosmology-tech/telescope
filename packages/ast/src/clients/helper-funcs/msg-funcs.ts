import * as ast from "@babel/types";
import { ProtoServiceMethod } from "@cosmology/types";
import { GenericParseContext } from "../../encoding";

/**
 *
 * @param context
 * @param service
 * @param methodKey e.g. "balance"
 * @param helperCreatorName e.g. "createGetBalance"
 * @returns
 */
export function createMsgHelperCreator(
    context: GenericParseContext,
    service: ProtoServiceMethod,
    helperCreatorName?: string
) {
    context.addUtil("buildTx");
    context.addUtil("ISigningClient");
    context.addUtil("SigningClientResolver");
    context.addUtil("toEncoders");
    context.addUtil("toConverters");

    const useGlobalDecoderRegistry =
        context.pluginValue("interfaces.enabled") &&
        context.pluginValue("interfaces.useGlobalDecoderRegistry") &&
        context.pluginValue("helperFuncCreators.enabled") &&
        context.pluginValue("helperFuncCreators.useGlobalDecoderRegistry");

    const callExpression = ast.callExpression(ast.identifier("buildTx"), [
        ast.objectExpression(
            [
                ast.objectProperty(
                    ast.identifier("clientResolver"),
                    ast.identifier("clientResolver"),
                    false,
                    true
                ),
                ast.objectProperty(
                    ast.identifier("typeUrl"),
                    ast.memberExpression(
                        ast.identifier(service.requestType),
                        ast.identifier("typeUrl")
                    )
                ),
                ast.objectProperty(
                    ast.identifier("encoders"),
                    ast.callExpression(ast.identifier("toEncoders"), [
                        ast.identifier(service.requestType),
                    ])
                ),
                ast.objectProperty(
                    ast.identifier("converters"),
                    ast.callExpression(ast.identifier("toConverters"), [
                        ast.identifier(service.requestType),
                    ])
                ),
                useGlobalDecoderRegistry &&
                    ast.objectProperty(
                        ast.identifier("deps"),
                        ast.arrayExpression([
                            ast.identifier(service.requestType),
                        ])
                    ),
            ].filter(Boolean)
        ),
    ]);
    callExpression.typeParameters = ast.tsTypeParameterInstantiation([
        ast.tsTypeReference(ast.identifier(service.requestType)),
    ]);
    const customHookArgumentsType = ast.tsTypeAnnotation(
        ast.tsTypeReference(ast.identifier("SigningClientResolver"))
    );
    const arg = ast.identifier("clientResolver");
    arg.typeAnnotation = customHookArgumentsType;
    arg.optional = true;
    const arrowFuncExp = ast.arrowFunctionExpression([arg], callExpression);
    return ast.exportNamedDeclaration(
        ast.variableDeclaration("const", [
            ast.variableDeclarator(
                ast.identifier(helperCreatorName),
                arrowFuncExp
            ),
        ])
    );
}

/**
 *
 * @param context
 * @param service
 * @param methodKey e.g. "balance"
 * @param helperCreatorName e.g. "createGetBalance"
 * @param hookName e.g. "useGetBalance"
 * @returns
 */
export function createMsgHooks(
    context: GenericParseContext,
    service: ProtoServiceMethod,
    helperCreatorName?: string,
    hookName?: string
) {
    context.addUtil("buildUseMutation");

    const callExpression = ast.callExpression(
        ast.identifier("buildUseMutation"),
        [
            ast.objectExpression([
                ast.objectProperty(
                    ast.identifier("builderMutationFn"),
                    ast.identifier(helperCreatorName)
                ),
            ]),
        ]
    );
    callExpression.typeParameters = ast.tsTypeParameterInstantiation([
        ast.tsTypeReference(ast.identifier(service.requestType)),
        ast.tsTypeReference(ast.identifier(`Error`)),
    ]);
    return ast.exportNamedDeclaration(
        ast.variableDeclaration("const", [
            ast.variableDeclarator(ast.identifier(hookName), callExpression),
        ])
    );
}
