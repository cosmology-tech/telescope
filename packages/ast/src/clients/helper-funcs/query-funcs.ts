import * as ast from "@babel/types";
import { ProtoServiceMethod } from "@cosmology/types";
import { GenericParseContext } from "../../encoding";

/**
 *
 * @param context
 * @param service
 * @param methodKey e.g. "balance"
 * @param helperCreatorName e.g. "createGetBalance"
 * @returns
 */
export function createQueryHelperCreator(
    context: GenericParseContext,
    service: ProtoServiceMethod,
    svcKey: string,
    methodKey?: string,
    helperCreatorName?: string
) {
    const pkgImportName = context.ref.proto.package + '.' + svcKey

    context.addUtil("RpcResolver");
    context.addUtil("buildQuery");
    const callExpression = ast.callExpression(ast.identifier("buildQuery"), [
        ast.objectExpression([
            ast.objectProperty(
                ast.identifier("encode"),
                ast.memberExpression(
                    ast.identifier(service.requestType),
                    ast.identifier("encode")
                )
            ),
            ast.objectProperty(
                ast.identifier("decode"),
                ast.memberExpression(
                    ast.identifier(service.responseType),
                    ast.identifier("decode")
                )
            ),
            ast.objectProperty(
                ast.identifier("service"),
                ast.stringLiteral(pkgImportName)
            ),
            ast.objectProperty(
                ast.identifier("method"),
                ast.stringLiteral(methodKey)
            ),
            ast.objectProperty(
                ast.identifier("clientResolver"),
                ast.identifier("clientResolver"),
                false, true
            ),
        ]),
    ]);
    callExpression.typeParameters = ast.tsTypeParameterInstantiation([
        ast.tsTypeReference(ast.identifier(service.requestType)),
        ast.tsTypeReference(ast.identifier(service.responseType)),
    ]);

    const customHookArgumentsType = ast.tsTypeAnnotation(
        ast.tsTypeReference(ast.identifier("RpcResolver"))
    );
    const arg = ast.identifier("clientResolver");
    arg.typeAnnotation = customHookArgumentsType;
    arg.optional = true;

    const arrowFuncExp = ast.arrowFunctionExpression([arg], callExpression);

    return ast.exportNamedDeclaration(
        ast.variableDeclaration("const", [
            ast.variableDeclarator(
                ast.identifier(helperCreatorName),
                arrowFuncExp
            ),
        ])
    );
}

/**
 *
 * @param context
 * @param service
 * @param methodKey e.g. "balance"
 * @param helperCreatorName e.g. "createGetBalance"
 * @param hookName e.g. "useGetBalance"
 * @returns
 */
export function createQueryHooks(
    context: GenericParseContext,
    service: ProtoServiceMethod,
    methodKey?: string,
    helperCreatorName?: string,
    hookName?: string
) {
    context.addUtil("buildUseQuery");
    const callExpression = ast.callExpression(ast.identifier("buildUseQuery"), [
        ast.objectExpression([
            ast.objectProperty(
                ast.identifier("builderQueryFn"),
                ast.identifier(helperCreatorName)
            ),
            ast.objectProperty(
                ast.identifier("queryKeyPrefix"),
                ast.stringLiteral(`${methodKey}Query`)
            ),
        ]),
    ]);
    callExpression.typeParameters = ast.tsTypeParameterInstantiation([
        ast.tsTypeReference(ast.identifier(service.requestType)),
        ast.tsTypeReference(ast.identifier(service.responseType)),
    ]);
    return ast.exportNamedDeclaration(
        ast.variableDeclaration("const", [
            ast.variableDeclarator(ast.identifier(hookName), callExpression),
        ])
    );
}
