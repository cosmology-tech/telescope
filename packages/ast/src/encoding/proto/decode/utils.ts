import * as t from '@babel/types';
import { DecodeMethod } from './index';
import { getInterfaceDecoderName, getKeyTypeEntryName } from '..';
import { TypeLong } from '../../../utils';

export const decode = {
    string(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.string(args));
    },
    bool(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.bool(args));
    },
    double(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.double(args));
    },
    float(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.float(args));
    },
    int32(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.int32(args));
    },
    sint32(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.sint32(args));
    },
    uint32(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.uint32(args));
    },
    fixed32(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.fixed32(args));
    },
    sfixed32(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.sfixed32(args));
    },
    int64(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.int64(args));
    },
    sint64(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.sint64(args));
    },
    uint64(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.uint64(args));
    },
    fixed64(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.fixed64(args));
    },
    sfixed64(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.sfixed64(args));
    },
    duration(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        const durationFormat = args.context.pluginValue('prototypes.typingsFormat.duration');

        switch (durationFormat) {
            case 'string':
                return switchOnTag(num, prop, baseTypes.durationString(args));
            case 'duration':
            default:
                return switchOnTag(num, prop, baseTypes.type(args));
        }
    },
    timestamp(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        const timestampFormat = args.context.pluginValue('prototypes.typingsFormat.timestamp')
        switch (timestampFormat) {
            case 'timestamp':
                return switchOnTag(num, prop, baseTypes.timestamp(args));
            case 'date':
            default:
                return switchOnTag(num, prop, baseTypes.timestampDate(args));
        }

    },
    type(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.type(args));
    },
    enum(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.enum(args));
    },
    bytes(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTag(num, prop, baseTypes.bytes(args));
    },
    keyHash(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchOnTagTakesArray(num, prop, baseTypes.keyHash(args));
    },
    tagDelimArray(args: DecodeMethod, expr: t.Expression) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchTagDelimArray(num,
            prop,
            expr
        )
    },
    array(args: DecodeMethod, expr: t.Expression) {
        const num = args.field.id;
        const prop = args.field.name;
        return switchArray(num,
            prop,
            expr
        )
    },
    typeArray(args: DecodeMethod) {
        const num = args.field.id;
        const prop = args.field.name;
        const name = args.context.getTypeName(args.field);

        if (
            !args.context.options.aminoEncoding.useLegacyInlineEncoding &&
            args.context.options.interfaces.enabled &&
            args.field.type === 'google.protobuf.Any' &&
            args.field.options['(cosmos_proto.accepts_interface)']

        ) {
            const interfaceName = args.field.options['(cosmos_proto.accepts_interface)'];
            const interfaceFnName = getInterfaceDecoderName(interfaceName)

            return switchAnyTypeArray(
                num,
                prop,
                name,
                interfaceFnName,
            );
        }
        return switchProtoTypeArray(
            args,
            num,
            prop,
            name
        );
    }

};

export const baseTypes = {

    // reader.string();
    string(args: DecodeMethod) {
        const useCosmosSDKDec = args.context.pluginValue(
          'prototypes.typingsFormat.customTypes.useCosmosSDKDec'
        );
        const isCosmosSDKDec =
            (args.field.options?.['(gogoproto.customtype)'] ==
                'github.com/cosmos/cosmos-sdk/types.Dec') ||
            (args.field.options?.['(gogoproto.customtype)'] ==
                'cosmossdk.io/math.LegacyDec');

        let valueExpression = t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('string')
            ),
            []
        );

        if(useCosmosSDKDec && isCosmosSDKDec){
          args.context.addUtil("Decimal");

          valueExpression = t.callExpression(
            t.memberExpression(
              t.callExpression(
                t.memberExpression(
                  t.identifier('Decimal'),
                  t.identifier('fromAtomics'),
                ),
                [
                  valueExpression,
                  t.numericLiteral(18)
                ]
              ),
              t.identifier('toString'),
            ),
            []
          )
        }

        return valueExpression;
    },

    // reader.bool();
    bool(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('bool')
            ),
            []
        );
    },

    // reader.double();
    double(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('double')
            ),
            []
        );
    },

    // reader.float();
    float(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('float')
            ),
            []
        );
    },

    // reader.int32();
    int32(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('int32')
            ),
            []
        );
    },

    // reader.sint32();
    sint32(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('sint32')
            ),
            []
        );
    },

    // reader.uint32();
    uint32(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('uint32')
            ),
            []
        );
    },

    // reader.fixed32();
    fixed32(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('fixed32')
            ),
            []
        );
    },

    // reader.sfixed32();
    sfixed32(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('sfixed32')
            ),
            []
        );
    },

    long(type: string, args: DecodeMethod) {
      TypeLong.addUtil(args.context);

      switch (TypeLong.getType(args.context)) {
        case 'BigInt':
          return t.callExpression(
            t.memberExpression(t.identifier('reader'), t.identifier(type)),
            []
          );

        case 'Long':
        default:
          return t.tsAsExpression(
              t.callExpression(
                  t.memberExpression(
                      t.identifier('reader'),
                      t.identifier(type)
                  ),
                  []
              ),
              t.tsTypeReference(
                  TypeLong.getPropIdentifier(args.context)
              )
          );
      }
    },

    // (reader.int64() as Long);
    int64(args: DecodeMethod) {
      return baseTypes.long('int64', args)
    },

    // (reader.sint64() as Long);
    sint64(args: DecodeMethod) {
      return baseTypes.long('sint64', args)
    },

    // (reader.uint64() as Long);
    uint64(args: DecodeMethod) {
      return baseTypes.long('uint64', args)
    },

    // (reader.fixed64() as Long);
    fixed64(args: DecodeMethod) {
      return baseTypes.long('fixed64', args)
    },

    // (reader.sfixed64() as Long);
    sfixed64(args: DecodeMethod) {
      return baseTypes.long('sfixed64', args)
    },

    // SignDocDirectAux.decode(reader, reader.uint32());
    protoType(args: DecodeMethod) {
        const name = args.context.getTypeName(args.field);
        return t.callExpression(
            t.memberExpression(
                t.identifier(name),
                t.identifier('decode')
            ),
            [
                t.identifier('reader'),
                t.callExpression(
                    t.memberExpression(
                        t.identifier('reader'),
                        t.identifier('uint32')
                    ),
                    []
                ),
                ...(args.context.options.interfaces.enabled ? [
                    t.identifier('useInterfaces'),
                ] : []),
            ]
        )
    },

    anyType(args: DecodeMethod) {
        // const { propName, origName } = getFieldNames(args.field);
        // const typeMap = args.context.store.getTypeUrlMap(args.context.ref);
        // console.log(JSON.stringify(typeMap, null, 2));
        // console.log(JSON.stringify(args.field, null, 2));
        const interfaceName = args.field.options['(cosmos_proto.accepts_interface)'];
        const interfaceFnName = getInterfaceDecoderName(interfaceName)

        return t.conditionalExpression(
            t.identifier('useInterfaces'),
            t.tsAsExpression(
                t.callExpression(
                    t.identifier(interfaceFnName),
                    [
                        t.identifier('reader')
                    ]
                ),
                t.tsTypeReference(
                    t.identifier('Any')
                )
            ),
            baseTypes.protoType(args)
        );
    },

    type(args: DecodeMethod) {
        if (
            !args.context.options.aminoEncoding.useLegacyInlineEncoding &&
            args.context.options.interfaces.enabled &&
            args.field.type === 'google.protobuf.Any' &&
            args.field.options['(cosmos_proto.accepts_interface)']

        ) {
            return baseTypes.anyType(args);
        }
        return baseTypes.protoType(args);
    },

    // (reader.int32() as any);
    enum(args: DecodeMethod) {
        return t.tsAsExpression(
            t.callExpression(
                t.memberExpression(
                    t.identifier('reader'),
                    t.identifier('int32')
                ),
                []
            ),
            t.tsAnyKeyword()
        )
    },

    // reader.bytes()
    bytes(args: DecodeMethod) {
        return t.callExpression(
            t.memberExpression(
                t.identifier('reader'),
                t.identifier('bytes')
            ),
            []
        );
    },

    // message.period = fromDuration(Duration.decode(reader, reader.uint32()));

    durationString(args: DecodeMethod) {
        args.context.addUtil('fromDuration');

        return t.callExpression(
            t.identifier('fromDuration'),
            [
                t.callExpression(
                    t.memberExpression(
                        t.identifier('Duration'),
                        t.identifier('decode')
                    ),
                    [
                        t.identifier('reader'),
                        t.callExpression(
                            t.memberExpression(
                                t.identifier('reader'),
                                t.identifier('uint32')
                            ),
                            []
                        )
                    ]
                )
            ]
        )
    },

    // message.periodReset = Timestamp.decode(reader, reader.uint32());

    timestamp(args: DecodeMethod) {

        return t.callExpression(
            t.memberExpression(
                t.identifier('Timestamp'),
                t.identifier('decode')
            ),
            [
                t.identifier('reader'),
                t.callExpression(
                    t.memberExpression(
                        t.identifier('reader'),
                        t.identifier('uint32')
                    ),
                    []
                )
            ]
        )
    },

    timestampDate(args: DecodeMethod) {

        args.context.addUtil('fromTimestamp');

        return t.callExpression(
            t.identifier('fromTimestamp'),
            [
                t.callExpression(
                    t.memberExpression(
                        t.identifier('Timestamp'),
                        t.identifier('decode')
                    ),
                    [
                        t.identifier('reader'),
                        t.callExpression(
                            t.memberExpression(
                                t.identifier('reader'),
                                t.identifier('uint32')
                            ),
                            []
                        )
                    ]
                )
            ]
        )
    },

    // const entry1 = LogEntry_LabelsEntry.decode(reader, reader.uint32());

    // if (entry1.value !== undefined) {
    //     message.labels[entry13.key] = entry13.value;
    // }

    keyHash(args: DecodeMethod) {
        const prop = args.field.name;
        const name = args.typeName;
        const id = args.field.id;
        const entryVariable = `entry${id}`;
        return [
            t.variableDeclaration(
                'const',
                [
                    t.variableDeclarator(
                        t.identifier(entryVariable),
                        t.callExpression(
                            t.memberExpression(
                                t.identifier(getKeyTypeEntryName(name, prop)),
                                t.identifier('decode')
                            ),
                            [
                                t.identifier('reader'),
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('reader'),
                                        t.identifier('uint32')
                                    ),
                                    []
                                )
                            ]
                        )
                    )
                ]
            ),
            t.ifStatement(
                t.binaryExpression(
                    '!==',
                    t.memberExpression(
                        t.identifier(entryVariable),
                        t.identifier('value')
                    ),
                    t.identifier('undefined')
                ),
                t.blockStatement([
                    t.expressionStatement(
                        t.assignmentExpression(
                            '=',
                            t.memberExpression(
                                t.memberExpression(
                                    t.identifier('message'),
                                    t.identifier(prop)
                                ),
                                t.memberExpression(
                                    t.identifier(entryVariable),
                                    t.identifier('key')
                                ),
                                true
                            ),
                            t.memberExpression(
                                t.identifier(entryVariable),
                                t.identifier('value')
                            )
                        )
                    )
                ])
            )
        ]
    }

};


export const switchOnTag = (num: number, prop: string, expr: t.Expression) => {
    return t.switchCase(
        t.numericLiteral(num),
        [
            t.expressionStatement(
                t.assignmentExpression(
                    '=',
                    t.memberExpression(
                        t.identifier('message'),
                        t.identifier(prop)
                    ),
                    expr
                )
            ),
            t.breakStatement()
        ]
    );
};

export const switchOnTagTakesArray = (num: number, prop: string, expr: t.Statement[]) => {
    return t.switchCase(
        t.numericLiteral(num),
        [
            ...expr,
            t.breakStatement()
        ]
    );
};

//    message.tokenInMaxs.push(Coin.decode(reader, reader.uint32()));
export const switchProtoTypeArray = (
    args: DecodeMethod,
    num: number,
    prop: string,
    name: string
) => {
    return t.switchCase(
        t.numericLiteral(num),
        [
            t.expressionStatement(
                t.callExpression(
                    t.memberExpression(
                        t.memberExpression(
                            t.identifier('message'),
                            t.identifier(prop)
                        ),
                        t.identifier('push')
                    ),
                    [
                        t.callExpression(
                            t.memberExpression(
                                t.identifier(name),
                                t.identifier('decode')
                            ),
                            [
                                t.identifier('reader'),
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('reader'),
                                        t.identifier('uint32')
                                    ),
                                    []
                                ),
                                ...(args.context.options.interfaces.enabled ? [
                                    t.identifier('useInterfaces'),
                                ] : []),
                            ]
                        )
                    ]
                )
            ),
            t.breakStatement()
        ]
    )
};

export const switchAnyTypeArray = (num: number, prop: string, typeName: string, interfaceName: string) => {
    return switchArray(num, prop,
        t.conditionalExpression(
            t.identifier('useInterfaces'),
            t.tsAsExpression(
                t.callExpression(
                    t.identifier(interfaceName),
                    [
                        t.identifier('reader')
                    ]
                ),
                t.tsTypeReference(
                    t.identifier('Any')
                )
            ),
            t.callExpression(
                t.memberExpression(
                    t.identifier(typeName),
                    t.identifier('decode')
                ),
                [
                    t.identifier('reader'),
                    t.callExpression(
                        t.memberExpression(
                            t.identifier('reader'),
                            t.identifier('uint32')
                        ),
                        []
                    ),
                    t.identifier('useInterfaces'),
                ]
            )
        )
    );
};

// if ((tag & 7) === 2) {
//     const end2 = reader.uint32() + reader.pos;

//     while (reader.pos < end2) {
//         message.codeIds.push((reader.uint64() as Long));
//     }
// } else {
//     message.codeIds.push((reader.uint64() as Long));
// }

export const switchTagDelimArray = (num: number, prop: string, expr: t.Expression) => {
    const blockStmt = t.blockStatement([
        t.expressionStatement(
            t.callExpression(
                t.memberExpression(
                    t.memberExpression(
                        t.identifier('message'),
                        t.identifier(prop)
                    ),
                    t.identifier('push')
                ),
                [
                    expr
                ]
            )
        )
    ]);
    return t.switchCase(
        t.numericLiteral(num),
        [
            t.ifStatement(
                t.binaryExpression(
                    '===',
                    t.binaryExpression(
                        '&',
                        t.identifier('tag'),
                        t.numericLiteral(7)
                    ),
                    t.numericLiteral(2)
                ),
                t.blockStatement([
                    t.variableDeclaration('const', [
                        t.variableDeclarator(
                            t.identifier('end2'),
                            t.binaryExpression(
                                '+',
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('reader'),
                                        t.identifier('uint32')
                                    ),
                                    []
                                ),
                                t.memberExpression(
                                    t.identifier('reader'),
                                    t.identifier('pos')
                                )
                            )
                        )
                    ]),
                    // while loop

                    t.whileStatement(
                        t.binaryExpression(
                            '<',
                            t.memberExpression(
                                t.identifier('reader'),
                                t.identifier('pos')
                            ),
                            t.identifier('end2')
                        ),
                        blockStmt
                    )

                ]),
                blockStmt
            ),
            t.breakStatement()
        ]
    )
};

export const switchArray = (num: number, prop: string, expr: t.Expression) => {
    return t.switchCase(
        t.numericLiteral(num),
        [
            t.expressionStatement(
                t.callExpression(
                    t.memberExpression(
                        t.memberExpression(
                            t.identifier('message'),
                            t.identifier(prop)
                        ),
                        t.identifier('push')
                    ),
                    [
                        expr
                    ]
                )
            ),
            t.breakStatement()
        ]
    );
};

