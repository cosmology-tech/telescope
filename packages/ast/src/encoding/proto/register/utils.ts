import * as t from "@babel/types";
import { ProtoType } from "@cosmology/types";
import { ProtoParseContext } from "../../context";
import { getAminoTypeName, getTypeUrl } from "@cosmology/utils";

export const createRegisterObject = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType
) => {
    if (name === "Any") {
        return;
    }

    context.addUtil("GlobalDecoderRegistry");

    const typeUrl = getTypeUrl(context.ref.proto, proto);
    if (!typeUrl) return;

    return t.expressionStatement(
        t.callExpression(
            t.memberExpression(
                t.identifier("GlobalDecoderRegistry"),
                t.identifier("register")
            ),
            [
                t.memberExpression(t.identifier(name), t.identifier("typeUrl")),
                t.identifier(name),
            ]
        )
    );
};

export const createRegisterAminoProtoMapping = (
    context: ProtoParseContext,
    name: string,
    proto: ProtoType
) => {
    if (name === "Any") {
        return;
    }
    context.addUtil("GlobalDecoderRegistry");

    const str = getAminoTypeName(context, context.ref.proto, proto);
    if (!str || str.startsWith("/")) return;

    return t.expressionStatement(
        t.callExpression(
            t.memberExpression(
                t.identifier("GlobalDecoderRegistry"),
                t.identifier("registerAminoProtoMapping")
            ),
            [
                t.memberExpression(
                    t.identifier(name),
                    t.identifier("aminoType")
                ),
                t.memberExpression(t.identifier(name), t.identifier("typeUrl")),
            ]
        )
    );
};
