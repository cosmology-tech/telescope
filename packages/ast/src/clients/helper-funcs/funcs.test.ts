import * as ast from "@babel/types";
import { GenericParseContext } from "../../encoding";
import { getNestedProto } from "@cosmology/utils";
import { traverse } from "@cosmology/proto-parser";
import { expectCode, getTestProtoStore, printCode } from "../../../test-utils";
import { defaultTelescopeOptions, ProtoService } from "@cosmology/types";
import { createMsgHelperCreator, createMsgHooks } from "./msg-funcs";
import { createQueryHelperCreator, createQueryHooks } from "./query-funcs";

const store = getTestProtoStore();
store.traverseAll();

it("query funcs works", async () => {
    const ref = store.findProto("cosmos/bank/v1beta1/query.proto");

    const res = traverse(store, ref);

    const { Query } = getNestedProto(res);

    const mockService: ProtoService = Query;

    const mockContext = new GenericParseContext(
        ref,
        store,
        defaultTelescopeOptions
    );

    expectCode(
        createQueryHelperCreator(
            mockContext,
            mockService.methods["Balance"],
            mockService.name,
            "Balance",
            "createGetBalance"
        )
    );

    expectCode(
        createQueryHooks(
            mockContext,
            mockService.methods["Balance"],
            "Balance",
            "createGetBalance",
            "useBalance"
        )
    );
});

it("msg funcs works", async () => {
    const ref = store.findProto("cosmos/bank/v1beta1/tx.proto");

    const res = traverse(store, ref);

    const { Msg } = getNestedProto(res);

    const mockService: ProtoService = Msg;

    const mockContext = new GenericParseContext(
        ref,
        store,
        defaultTelescopeOptions
    );

    expectCode(
        createMsgHelperCreator(
            mockContext,
            mockService.methods["Send"],
            "createSend"
        )
    );

    expectCode(
        createMsgHooks(
            mockContext,
            mockService.methods["Send"],
            "createSend",
            "useSend"
        )
    );
});
