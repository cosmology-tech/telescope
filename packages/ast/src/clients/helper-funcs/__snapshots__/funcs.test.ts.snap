// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`msg funcs works 1`] = `
"export const createSend = (getSigningClient: SigningClientResolver) => buildTx<MsgSend>({
  getSigningClient: getSigningClient,
  typeUrl: MsgSend.typeUrl,
  encoders: toEncoders(MsgSend),
  converters: toConverters(MsgSend)
});"
`;

exports[`msg funcs works 2`] = `
"export const useSend = buildUseMutation<MsgSend, Error>({
  builderMutationFn: createSend
});"
`;

exports[`query funcs works 1`] = `
"export const createGetBalance = (getRpcInstance: SigningClientResolver) => buildQuery<QueryBalanceRequest, QueryBalanceResponse>({
  encoder: QueryBalanceRequest.encode,
  decoder: QueryBalanceResponse.decode,
  service: "cosmos.bank.v1beta1.Query",
  method: "Balance",
  getRpcInstance: getRpcInstance
});"
`;

exports[`query funcs works 2`] = `
"export const useBalance = buildUseQuery<QueryBalanceRequest, QueryBalanceResponse>({
  builderQueryFn: createGetBalance,
  queryKeyPrefix: "BalanceQuery"
});"
`;
